sormas-backend/src/main/java/de/symeda/sormas/backend/contact/ContactFacadeEjb.java

```diff
@@ -1600,4 +1624,244 @@ public boolean isContactEditAllowed(String contactUuid) {
+			completeness += 0.2f;
+		}
+		if (sampleService.getSampleCountByContact(contact) > 0) {
+			completeness += 0.10f;
+		}
+		if (contact.getVisits().size() > 0) {
+			completeness += 0.10f;
+		}
+		if (contact.getPerson().getBirthdateYYYY() != null || contact.getPerson().getApproximateAge() != null) {
+			completeness += 0.05f;
+		}
+		if (contact.getPerson().getSex() != null) {
+			completeness += 0.05f;
+		}
+		if (contact.getCaseOrEventInformation() != null) {
+			completeness += 0.05f;
+		}
+		if (contact.getRelationDescription() != null) {
+			completeness += 0.05f;
+		}
+		return completeness;
+	}
+
+	@Override
+	public void mergeContact(String leadUuid, String otherUuid) {
+		mergeContact(getContactByUuid(leadUuid), getContactByUuid(otherUuid), false);
+	}
+
+	private void mergeContact(ContactDto leadContact, ContactDto otherContact, boolean cloning) {
+		// 1 Merge Dtos
+		// 1.1 Contact
+		DtoHelper.fillDto(leadContact, otherContact, cloning);
+		saveContact(leadContact, !cloning, false);
+
+		// 1.2 Person
+		if (!cloning) {
+			PersonDto leadPerson = personFacade.getPersonByUuid(leadContact.getPerson().getUuid());
+			PersonDto otherPerson = personFacade.getPersonByUuid(otherContact.getPerson().getUuid());
+			DtoHelper.fillDto(leadPerson, otherPerson, cloning);
+			personFacade.savePerson(leadPerson);
+		} else {
+			assert (DataHelper.equal(leadContact.getPerson().getUuid(), otherContact.getPerson().getUuid()));
+		}
+
+		// 2 Change ContactReference
+		Contact leadCont = contactService.getByUuid(leadContact.getUuid());
+		Contact otherCont = contactService.getByUuid(otherContact.getUuid());
+
+		// 2.2 Samples
+		List<Sample> samples = sampleService.findBy(new SampleCriteria().contact(otherCont.toReference()), null);
+		for (Sample sample : samples) {
+			if (cloning) {
+				SampleDto newSample = SampleDto.build(sample.getReportingUser().toReference(), leadCont.toReference());
+				DtoHelper.fillDto(newSample, SampleFacadeEjb.toDto(sample), cloning);
+				sampleFacade.saveSample(newSample, false);
+
+				// 2.2.1 Pathogen Tests
+				for (PathogenTest pathogenTest : sample.getPathogenTests()) {
+					PathogenTestDto newPathogenTest = PathogenTestDto.build(newSample.toReference(), pathogenTest.getLabUser().toReference());
+					DtoHelper.fillDto(newPathogenTest, PathogenTestFacadeEjb.PathogenTestFacadeEjbLocal.toDto(pathogenTest), cloning);
+					sampleTestFacade.savePathogenTest(newPathogenTest);
+
+				}
+			} else {
+				// simply move existing entities to the merge target
+				sample.setAssociatedContact(leadCont);
+				sampleService.ensurePersisted(sample);
+			}
+		}
+
+		// 2.4 Tasks
+		if (!cloning) {
+			// simply move existing entities to the merge target
+
+			List<Task> tasks = taskService.findBy(new TaskCriteria().contact(new ContactReferenceDto(otherCont.getUuid())), true);
+			for (Task task : tasks) {
+				task.setContact(leadCont);
+				taskService.ensurePersisted(task);
+			}
+		}
```

Comment:
The case merging also puts visits from otherCase to leadCase to make sure that no data is lost; please also implement this for contacts (see CaseFacadeEjb:3097 on the dev branch)